### PR TITLE
Add TypeFilter enabled benchmark

### PR DESCRIPTION
**Pull request by Arkatufus**
Add TypeFilter benchmark tests

### CODE REVIEW

File: src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs [hunk 1]
```diff
@@ -20,9 +21,25 @@ public class SerializeClassesBenchmark : HyperionBenchmark
         private LargeSealedClass sealedObject;
         private GenericClass<int, string, bool, DateTime, Guid> genericObject;
 
+        private Serializer _filteredSerializer;
+        
         protected override void Init()
         {
-            Serializer = new Serializer(new SerializerOptions(preserveObjectReferences:true));
+            var baseOptions = new SerializerOptions(preserveObjectReferences: true);
+            Serializer = new Serializer(baseOptions);
```
Comment (eaba) [inline src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs:29]: Is this serving a different purpose compared to `_filteredSerializer`? if not, then we can do this `Serializer = _filteredSerializer;`.

File: src/Hyperion.Benchmarks/SerializeStructsBenchmark.cs [hunk 2]
```diff
@@ -30,8 +43,21 @@ protected override void Init()
...
+        [Benchmark] public void Filtered_Enums() => SerializeAndFilteredDeserialize(testEnum);
         [Benchmark] public void Standard_Value_Types() => SerializeAndDeserialize(standardValue);
+        [Benchmark] public void Filtered_Standard_Value_Types() => SerializeAndFilteredDeserialize(standardValue);
         [Benchmark] public void Blittable_Value_Types() => SerializeAndDeserialize(blittableValue);
+        [Benchmark] public void Filtered_Blittable_Value_Types() => SerializeAndFilteredDeserialize(blittableValue);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SerializeAndFilteredDeserialize<T>(T elem)
+        {
+            Serializer.Serialize(elem, Stream);
```
Comment (eaba) [inline src/Hyperion.Benchmarks/SerializeStructsBenchmark.cs:55]: `Serializer` does not have types registered, it should throw an exception. How about `_filteredSerializer.Serialize(elem, Stream)`?;

File: src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs [hunk 3]
```diff
@@ -45,9 +62,23 @@ protected override void Init()
...
+        [Benchmark] public void Filtered_Virtual_Classes() => SerializeAndFilteredDeserialize(virtualObject);
         [Benchmark] public void Large_Sealed_Classes() => SerializeAndDeserialize(sealedObject);
+        [Benchmark] public void Filtered_Large_Sealed_Classes() => SerializeAndFilteredDeserialize(sealedObject);
         [Benchmark] public void Generic_Classes() => SerializeAndDeserialize(genericObject);
+        [Benchmark] public void Filtered_Generic_Classes() => SerializeAndFilteredDeserialize(genericObject);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SerializeAndFilteredDeserialize<T>(T elem)
+        {
+            Serializer.Serialize(elem, Stream);
```
Comment (eaba) [inline src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs:76]: `Serializer` does not have types registered, it should throw an exception. How about `_filteredSerializer.Serialize(elem, Stream)`?;

File: src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs:29]: Yeah that's an interesting question - why do we have two different serializer instances here for these benchmarks?

Comment (Arkatufus) [reply src/Hyperion.Benchmarks/SerializeClassesBenchmark.cs:29]: `Serializer` is the one that came with the benchmark spec, its the one with the type filtering disabled, we're benchmarking 2 different serializer, one disabled and one with type filtering turned on.

File: src/Hyperion.Benchmarks/SerializeStructsBenchmark.cs (hunk 2 above)
Comment (Arkatufus) [reply src/Hyperion.Benchmarks/SerializeStructsBenchmark.cs:55]: if you don't have any types registered, it automatically disable the type filtering. we don't want to force our users to always use type filtering, its an opt-in feature.

Comment (Arkatufus) [reply src/Hyperion.Benchmarks/SerializeStructsBenchmark.cs:55]: Type filtering is only done during __deserialization__, it doesn't matter who serializes the original object, the deserializer is the one responsible for filtering the types.

Comment (eaba) [reply src/Hyperion.Benchmarks/SerializeStructsBenchmark.cs:55]: Got that - I forgot that part! :)

**Comment by eaba** [comment]
LGTM

**Pull request merged**